proposals/ytt/005-semver/README.md

```diff
@@ -0,0 +1,217 @@
+```
+
+#### version·cmp
+
+`version.cmp(other)` compares two semantic versions and returns `0` if they are equal, `-1` if `other` is greater
+and `1` if other is smaller. If `other` is not a `version` struct, it is a dynamic error.
+
+Usage of `version.(eq, lt, lte, gt, gte)` is recommended instead. However, there may be cases where a numeric return
+value is preferred.
+
+```yaml
+#@ load("@ytt:semver", "semver")
+
+#@ semver.version().cmp(semver.version())                  #! 0
+#@ semver.version(1).cmp(semver.version(2))                #! -1
+#@ semver.version(1, 2, 3).cmp(semver.version(2, 0, 3))    #! 1
+#@ semver.version(1, 2, 3).cmp(dict())                     #! ⚡️ error
+```
+
+#### version·(eq, lt, lte, gt, gte)
+
+All of `version.eq(version)`, `version.lt(version)`, `version.lte(version)`, `version.gt(version)`
+and `version.gte(version)` compare `version` with `other` and return `True` if the operator holds and `False` otherwise.
+If `other` is not a `version` struct, it is a dynamic error.
+
+```yaml
+#@ load("@ytt:data", "data")
+#@ load("@ytt:semver", "semver")
+#@ load("@ytt:assert", "assert")
+
+#@ v1_2_3 = semver.version(1, 2, 3)
+#@ given_version = semver.from_str(data.values.version)
+
+---
+cool_config: nice
+#@ if/end given_version.lt(v1_2_3)
+ye_olde_confyg: they don't make 'em like this any more
+
+#@ if given_version.gte(semver.version(10, 20, 30))
+#@   assert.fail("Sorry, but that's too hot for us")
+#@ end
+```
+
+#### version·to_str
+
+`version.to_str()` returns the string representation of the `version` struct.
+
+```yaml
+#@ load("@ytt:semver", "semver")
+
+#@ semver.version().to_str()                               #! "0.0.0"
+#@ semver.version(1, 2, 3).to_str()                        #! "1.2.3"
+#@ semver.version(1, 2, 3, "beta").to_str()                #! "1.2.3-beta"
+#@ semver.version(1, 2, 3, metadata="build.5").to_str()    #! "1.2.3+build.5"
+#@ semver.version(1, 2, 3, "next", "nightly").to_str()     #! "1.2.3-next+nightly"
+#@ semver.version(minor=1, prerelease="alpha").to_str()    #! "0.1.0-alpha"
+```
+
+Use cases:
+
+```yaml
+#@ load("@ytt:semver", "semver")
+
+#@ all_the_minor_versions = [semver.version(minor=i).to_str() for i in range(10)]
+---
+allowed_versions: #@ all_the_minor_versions
+```
+
+### Other Approaches Considered
+
+* A Github-hosted Starlark library like `awesome-ytt/semver` containing a similar API could be fetched with `vendir` and
+  then loaded. This wouldn't in reach for the widest possible spectrum of configuration authors. It would require to
+  reimplement what existing, established Go modules already have done.
+
+## Open Questions
+
+* There's a host of conceivable additional operations available for semver, e.g. test against range, sort list of
+  versions, bumping, etc. It is debatable if that should be part of the standard library. Testing against a range and
+  sorting are likelier candidates for a standard library. Bumping is possible with the primitives presented.
```

Comment:
One solid data point for validating absorbing logic like this is that it already exists. Having examples of these in hand not only provide great use cases, they also can help steer questions of scope like this one.

What custom code have configuration authors already written? (and I'm thinking about other systems as well: Helm chart templating, CUE libs, etc...

Reply:
> One solid data point for validating absorbing logic like this is that it already exists.

Cool! I read this as _"consider extending the proposal with sorting, range tests and modifications"_. Am I right?

> Having examples of these in hand not only provide great use cases, they also can help steer questions of scope like this one.

In other words, you are saying _"find data, act on it"_, aren't you?